File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

vector<int> answer = {0, 0};

void check(int startRow, int startCol, int size, vector<vector<int>>& arr){
    if(size == 1){
        answer[arr[startRow][startCol]]++;
        return;
    }
    
    int num = arr[startRow][startCol];
    for(int r = startRow; r < startRow + size; r++){
        for(int c = startCol; c < startCol + size; c++){
            if(arr[r][c] == num) continue;
            
            check(startRow, startCol, size / 2, arr);
            check(startRow, startCol + size / 2, size / 2, arr);
            check(startRow + size / 2, startCol, size / 2, arr);
            check(startRow + size / 2, startCol + size / 2, size / 2, arr);
            return;
        }
    }
    
    answer[num]++;
}

vector<int> solution(vector<vector<int>> arr) {
    check(0, 0, arr.size(), arr);   
    
    return answer;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    answer = [0, 0]
    
    def check(row, col, size):
        if size == 1:
            answer[arr[row][col]] += 1
            return
        
        for r in range(row, row + size):
            for c in range(col, col + size):
                if arr[r][c] != arr[row][col]:
                    check(row, col, size // 2)
                    check(row, col + size // 2, size // 2)
                    check(row + size // 2, col, size // 2)
                    check(row + size // 2, col + size // 2, size // 2)
                    return
        answer[arr[row][col]] += 1
    
    check(0, 0, len(arr))
    
    return answer